homeassistant/components/miele/services.py

```diff
@@ -0,0 +1,85 @@
+"""Services for Miele integration."""
+
+import logging
+from typing import cast
+
+import aiohttp
+import voluptuous as vol
+
+from homeassistant.const import CONF_DEVICE_ID
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.exceptions import HomeAssistantError, ServiceValidationError
+from homeassistant.helpers import config_validation as cv, device_registry as dr
+from homeassistant.helpers.service import async_extract_config_entry_ids
+
+from .const import DOMAIN, PROGRAM_ID
+from .coordinator import MieleConfigEntry
+
+SERVICE_PROGRAM = cv.make_entity_service_schema(
+    {
+        vol.Required("program_id"): cv.positive_int,
+        vol.Optional("duration"): cv.positive_int,
+        vol.Optional("temperature"): cv.positive_int,
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def extract_our_config_entry_ids(service_call: ServiceCall) -> list[str]:
+    """Extract config entry IDs from the service call."""
+    hass = service_call.hass
+    return [
+        entry_id
+        for entry_id in await async_extract_config_entry_ids(hass, service_call)
+        if (entry := hass.config_entries.async_get_entry(entry_id))
+        and entry.domain == DOMAIN
+    ]
```

Comment:
I think this would be cleaner to combine it with `async_loaded_entries`, and return directly a MieleConfigEntry:
```python
    target_entry_ids = await async_extract_config_entry_ids(hass, service_call)
    target_entries: list[MieleConfigEntry] = [
        loaded_entry
        for loaded_entry in hass.config_entries.async_loaded_entries(DOMAIN)
        if loaded_entry.entry_id in target_entry_ids
    ]
    if not target_entries:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="set_program_no_device",
        )
    return target_entries[0]
```